cli/internal/globby/globby.go

```diff
@@ -1,55 +1,141 @@
-		exclude = append(exclude, filepath.Join(basePath, p))
+	return relativePath, nil
+}
+
+// globFilesFs searches the specified file system to ensure to enumerate all files to include.
+func globFilesFs(fs afero.IOFS, basePath string, includePatterns []string, excludePatterns []string) []string {
+	var processedIncludes []string
+	var processedExcludes []string
+	var result = &set{
+		storageMap: map[string]bool{},
 	}
 
-	includePattern := "{" + strings.Join(include, ",") + "}"
-	excludePattern := "{" + strings.Join(exclude, ",") + "}"
+	for _, includePattern := range includePatterns {
+		includePath := filepath.Join(basePath, includePattern)
+		relativePath, err := getRelativePath(basePath, includePath)
+
+		if err == nil {
+			// Includes only operate on files.
+			processedIncludes = append(processedIncludes, relativePath)
+		}
+	}
 
-	_ = afero.Walk(fs, basePath, func(path string, info os.FileInfo, err error) error {
-		isDir := info.IsDir()
-		if val, _ := doublestar.PathMatch(excludePattern, path); val {
-			if isDir {
-				return filepath.SkipDir
-			}
+	for _, excludePattern := range excludePatterns {
+		excludePath := filepath.Join(basePath, excludePattern)
+		relativePath, err := getRelativePath(basePath, excludePath)
+
+		if err == nil {
+			// Excludes operate on entire folders.
+			processedExcludes = append(processedExcludes, filepath.Join(relativePath, "**"))
+		}
+	}
+
+	// We start from a naive includePattern
+	includePattern := ""
+	includeCount := len(processedIncludes)
+
+	// Do not use alternation if unnecessary.
+	if includeCount == 1 {
+		includePattern = filepath.Join(basePath, processedIncludes[0])
+	} else if includeCount > 1 {
+		// We start from a basePath prefix which allows doublestar to optimize the access.
+		includePattern = filepath.Join(basePath, "{"+strings.Join(processedIncludes, ",")+"}")
+	}
+
+	// We only create an exclude pattern if we have excludes.
+	var excludePattern string
+	excludeCount := len(processedExcludes)
+
+	// Do not use alternation if unnecessary.
+	if excludeCount == 1 {
+		excludePattern = filepath.Join(basePath, processedExcludes[0])
+	} else if excludeCount > 1 {
+		// We start from a basePath prefix which allows doublestar to optimize the access.
+		excludePattern = filepath.Join(basePath, "{"+strings.Join(processedExcludes, ",")+"}")
+	}
+
+	err := doublestar.GlobWalk(fs, filepath.ToSlash(includePattern), func(path string, dirEntry iofs.DirEntry) error {
+		// Root paths do not prepend the leading slash.
+		if !strings.HasPrefix(path, "/") {
+			path = filepath.Join(basePath, path)
+		}
+
+		if dirEntry.IsDir() {
 			return nil
 		}
 
-		if isDir {
+		if excludeCount == 0 {
+			result.Add(path)
 			return nil
 		}
 
-		if val, _ := doublestar.PathMatch(includePattern, path); val || len(includePatterns) == 0 {
-			result = append(result, path)
+		isExcluded, err := doublestar.Match(filepath.ToSlash(excludePattern), filepath.ToSlash(path))
+
+		if err == nil && !isExcluded {
```

Comment:
What does an error mean here? If it's a bad exclusion glob, we ought to report it

Reply:
Yup, the only thing that `err` can be here is `path.ErrBadPattern`. I'll capture this in the follow-up PR as well.